src/builder/QuarkBuilderBase.sol

```diff
@@ -0,0 +1,642 @@
+import {TokenWrapper} from "src/builder/TokenWrapper.sol";
+import {QuarkOperationHelper} from "src/builder/QuarkOperationHelper.sol";
+import {List} from "src/builder/List.sol";
+
+contract QuarkBuilderBase {
+    /* ===== Output Types ===== */
+
+    struct BuilderResult {
+        // Version of the builder interface. (Same as VERSION, but attached to the output.)
+        string version;
+        // Array of quark operations to execute to fulfill the client intent
+        IQuarkWallet.QuarkOperation[] quarkOperations;
+        // Array of action context and other metadata corresponding 1:1 with quarkOperations
+        Actions.Action[] actions;
+        // Struct containing containing EIP-712 data for a QuarkOperation or MultiQuarkOperation
+        EIP712Helper.EIP712Data eip712Data;
+        // Client-provided paymentCurrency string that was used to derive token addresses.
+        // Client may re-use this string to construct a request that simulates the transaction.
+        string paymentCurrency;
+    }
+
+    /* ===== Constants ===== */
+
+    string constant VERSION = "0.1.2";
+
+    /* ===== Custom Errors ===== */
+
+    error AssetPositionNotFound();
+    error FundsUnavailable(string assetSymbol, uint256 requiredAmount, uint256 actualAmount);
+    error InvalidActionChain();
+    error InvalidActionType();
+    error InvalidInput();
+    error MaxCostTooHigh();
+    error MissingWrapperCounterpart();
+    error InvalidRepayActionContext();
+
+    struct ActionIntent {
+        address actor;
+        uint256[] amountOuts;
+        string[] assetSymbolOuts;
+        uint256[] amountIns;
+        string[] assetSymbolIns;
+        uint256 blockTimestamp;
+        uint256 chainId;
+        bool useQuotecall;
+        bool bridgeEnabled;
+        bool autoWrapperEnabled;
+    }
+
+    /**
+     * @dev Collects assets for an action by checking and bridging assets if necessary to accomodate the intended action.
+     */
+    function collectAssetsForAction(
+        ActionIntent memory actionIntent,
+        Accounts.ChainAccounts[] memory chainAccountsList,
+        PaymentInfo.Payment memory payment,
+        IQuarkWallet.QuarkOperation memory actionQuarkOperation,
+        Actions.Action memory action
+    )
+        internal
+        pure
+        returns (IQuarkWallet.QuarkOperation[] memory quarkOperationsArray, Actions.Action[] memory actionsArray)
+    {
+        // Sanity check on ActionIntent
+        if (
+            actionIntent.amountOuts.length != actionIntent.assetSymbolOuts.length
+                || actionIntent.amountIns.length != actionIntent.assetSymbolIns.length
+        ) {
+            revert InvalidInput();
+        }
+
+        List.DynamicArray memory actions = List.newList();
+        List.DynamicArray memory quarkOperations = List.newList();
+
+        // Flag to check if the assetSymbolOut(used/supplied/transferred out) is the same as the payment token
+        bool paymentTokenInAssetSymbolOut = false;
```

Comment:
Should be `paymentTokenIsAssetSymbolOut`

Reply:
Oh I see.  I think I'm using In, it's becuase there is a array of symbolOut, and this flag will be true if one of them is payment token. 😄

Reply:
Is `paymentTokenIsPartOfAssetSymbolOuts` sounds better?